zeus/metric.py

```diff
@@ -0,0 +1,532 @@
+                    ["window", "index"],
+                    buckets=self.cpu_bucket_range,
+                    registry=self.registry,
+                )
+            # Initialize CPU and DRAM histograms
+            # Only when CPUs are available, we check if DRAM is available.
+            for i, cpu in enumerate(get_cpus().cpus):
+                if cpu.supportsGetDramEnergyConsumption():
+                    self.dram_histograms[i] = Histogram(
+                        f"energy_monitor_dram_{i}_energy_joules",
+                        f"DRAM {i} energy consumption",
+                        ["window", "index"],
+                        buckets=self.dram_bucket_range,
+                        registry=self.registry,
+                    )
+
+        self.max_gpu_bucket = max(self.gpu_bucket_range)
+        self.max_cpu_bucket = max(self.cpu_bucket_range)
+        self.max_dram_bucket = max(self.dram_bucket_range)
+
+        self.energy_monitor = ZeusMonitor(
+            cpu_indices=cpu_indices, gpu_indices=gpu_indices
+        )
+
+    def begin_window(self, name: str) -> None:
+        """Begin the energy monitoring window.
+
+        Args:
+            name (str): The unique name of the measurement window. Must match between calls to 'begin_window' and 'end_window'.
+        """
+        self.energy_monitor.begin_window(
+            f"__EnergyHistogram_{name}", sync_execution=True
+        )
+
+    def end_window(self, name: str) -> None:
+        """End the current energy monitoring window and record the energy data.
+
+        Retrieves the energy consumption data (for GPUs, CPUs, and DRAMs) for the monitoring window
+        and updates the corresponding Histogram metrics. The data is then pushed to the Prometheus Push Gateway.
+
+        Args:
+            name (str): The unique name of the measurement window. Must match between calls to 'begin_window' and 'end_window'.
+
+        Pushes:
+            - GPU energy data to the Prometheus Push Gateway via the associated Histogram metric.
+            - CPU energy data to the Prometheus Push Gateway via the associated Histogram metric.
+            - DRAM energy data to the Prometheus Push Gateway via the associated Histogram metric.
+        """
+        measurement = self.energy_monitor.end_window(
+            f"__EnergyHistogram_{name}", sync_execution=True
+        )
+
+        if measurement.gpu_energy:
+            for gpu_index, gpu_energy in measurement.gpu_energy.items():
+                if gpu_index in self.gpu_histograms:
+                    self.gpu_histograms[gpu_index].labels(
+                        window=name, index=gpu_index
+                    ).observe(gpu_energy)
+                if gpu_energy > self.max_gpu_bucket:
+                    warnings.warn(
+                        f"GPU {gpu_index} energy {gpu_energy} exceeds the maximum bucket value of {self.max_gpu_bucket}",
+                        stacklevel=1,
+                    )
```

Comment:
There should be a warning when the measured energy is smaller than the minimum bucket as well.